doc/src/guides/custom-functions.md

```diff
@@ -0,0 +1,1850 @@
+{class}`~.Function`. The name of the subclass will be the name of the
+function. Different methods should then be defined on this subclass, depending
+on what functionality you want to provide.
+
+As a motivating example for this document, let's create a custom function
+class representing the [versine
+function](https://en.wikipedia.org/wiki/Versine). Versine is a trigonometric
+function which was used historically alongside some of the more familiar
+trigonometric functions like sine and cosine. It is rarely used today. Versine
+can be defined by the identity
+
+(custom-functions-versine-definition)=
+
+$$\operatorname{versin}(x) = 1 - \cos(x).$$
+
+SymPy does not already include versine because it is used so rarely in modern
+mathematics and because it is so easily defined in terms of the more familiar
+cosine.
+
+Let us start by subclassing `Function`.
+
+```py
+>>> class versin(Function):
+...     pass
+```
+
+At this point, `versin` has no behaviors defined on it. It is very similar to
+the [undefined functions](custom-functions-fully-symbolic) we discussed above.
+Note that `versin` is a class, and `versin(x)` is an instance of this class.
+
+```py
+>>> versin(x)
+versin(x)
+>>> isinstance(versin(x), versin)
+True
+```
+
+```{note}
+
+All the methods described below are optional. They can be included if you want
+to define the given behavior, but if they are omitted, SymPy will default to
+leaving things unevaluated. For example, if you do not define
+[differentiation](custom-functions-differentiation), {func}`~.diff` will just
+return an unevaluated {class}`~.Derivative`.
+
+```
+
+(custom-functions-eval)=
+### Defining Automatic Evaluation with `eval()`
+
+```{sidebar} Reminder
+Remember that `eval()` should be defined with the `@classmethod` decorator.
+```
+
+The first and most common thing we might want to define on our custom function
+is automatic evaluation, that is, the cases where it will return an actual
+value instead of just remaining unevaluated as-is.
+
+This is done by defining the class method `eval()`. `eval()` should take the
+arguments of the function and return either a value or `None`. If it returns
+`None`, the function will remain unevaluated in that case. This also serves to
+define the signature of the function (by default, without an `eval()` method, a
+`Function` subclass will accept any number of arguments).
+
+For our function `versin`, we might recall that $\cos(n\pi) = (-1)^n$ for
+integer $n$, so $\operatorname{versin}(n\pi) = 1 - (-1)^n.$ We can make
+`versin` automatically evaluate to this value when passed an integer multiple
+of `pi`:
+
+(custom-functions-versin-eval-example)=
+```py
+>>> from sympy import pi, Integer
+>>> class versin(Function):
+...    @classmethod
+...    def eval(cls, x):
+...        # If x is an integer multiple of pi, x/pi will cancel and be an Integer
+...        n = x/pi
+...        if isinstance(n, Integer):
+...            return 1 - (-1)**n
```

Comment:
Maybe it's worth discussing the use of sympify somewhere. I think you've changed it so that both the inputs and outputs of this function are automatically sympified. That's very different from other Basic subclasses so it should probably be clarified somewhere.

Reply:
With things being sympified automatically it's not something you really need to worry about. I'd love for all `Basic` subclasses and methods to work like that, and also for all classes to more generally use an `eval` pattern rather than overriding `__new__`. There are too many bugs caused by people forgetting to sympify inputs/outputs, plus people tend to use sympify instead of _sympify when they do it themselves.